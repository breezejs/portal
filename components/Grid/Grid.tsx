import React, {memo, ReactNode} from 'react';
import styled from 'styled-components';
import {GlobalStylesTheme} from '../../core/GlobalStyles';

const Container = styled.div`
  display: grid;
  grid-template-columns: repeat(${
    ({columns}: IGridProps) => columns || GlobalStylesTheme.gridColumns
  };, [col-start] 1fr);
  grid-gap: ${({gutter}: IGridProps) => gutter || GlobalStylesTheme.gridGutter};
`;

interface IGridProps {
  children?: ReactNode;
  columns?: number;
  gutter?: string;
}

function Grid ({children, columns, gutter}: IGridProps) {
  return (
    <Container columns={columns} gutter={gutter}>{children}</Container>
  );
}

export default memo(Grid);
