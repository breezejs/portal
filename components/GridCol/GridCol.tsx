import React, {memo, ReactNode} from 'react';
import styled from 'styled-components';

const Col = styled.div`
  grid-column: ${({xs}: IGridColProps) => `span ${xs}`};
  position: relative;

  ${({sm}: IGridColProps) => sm ? `
    @media (min-width: 576px) {
      grid-column: span ${sm};
    }
  ` : null}

  ${({md}: IGridColProps) => md ? `
    @media (min-width: 768px) {
      grid-column: span ${md};
    }
  ` : null}

  ${({lg}: IGridColProps) => lg ? `
    @media (min-width: 992px) {
      grid-column: span ${lg};
    }
  ` : null}
`;

type GridColType =
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12;

interface IGridColProps {
  children?: ReactNode;
  lg?: GridColType;
  md?: GridColType;
  sm?: GridColType;
  xs?: GridColType;
}

function GridCol ({children, lg, md, sm, xs = 12}: IGridColProps) {
  return (
    <Col lg={lg} md={md} sm={sm} xs={xs}>{children}</Col>
  );
}

export default memo(GridCol);
