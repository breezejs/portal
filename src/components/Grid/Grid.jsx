import PropTypes from 'prop-types';
import React, {memo} from 'react';
import styled from 'styled-components';
import {GlobalStylesTheme} from '../GlobalStyles';

const Container = styled.div`
  display: grid;
  grid-template-columns: repeat(${({columns}) => columns}, [col-start] 1fr);
  grid-gap: ${({gutter}) => gutter};
`;

function Grid ({children, columns = GlobalStylesTheme.gridColumns, gutter = GlobalStylesTheme.gridGutter}) {
  return (
    <Container columns={columns} gutter={gutter}>{children}</Container>
  );
}

Grid.propTypes = {
  children: PropTypes.node,
  columns: PropTypes.number,
  gutter: PropTypes.string
};

export default memo(Grid);
