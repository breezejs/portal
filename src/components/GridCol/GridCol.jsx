import PropTypes from 'prop-types';
import React, {memo} from 'react';
import styled from 'styled-components';

const Col = styled.div`
  grid-column: ${({xs}) => `span ${xs}`};
  position: relative;

  ${({sm}) => sm ? `
    @media (min-width: 576px) {
      grid-column: span ${sm};
    }
  ` : null}

  ${({md}) => md ? `
    @media (min-width: 768px) {
      grid-column: span ${md};
    }
  ` : null}

  ${({lg}) => lg ? `
    @media (min-width: 992px) {
      grid-column: span ${lg};
    }
  ` : null}
`;

function GridCol ({children, lg, md, sm, xs}) {
  return (
    <Col lg={lg} md={md} sm={sm} xs={xs}>{children}</Col>
  );
}

GridCol.propTypes = {
  children: PropTypes.node,
  lg: PropTypes.number,
  md: PropTypes.number,
  sm: PropTypes.number,
  xs: PropTypes.number
};

GridCol.defaultProps = {
  children: null,
  lg: null,
  md: null,
  sm: null,
  xs: 12
};

export default memo(GridCol);
