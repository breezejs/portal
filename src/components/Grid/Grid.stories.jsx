import {number, text, withKnobs} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import React from 'react';
import GlobalStyles from '../GlobalStyles';
import GridCol from '../GridCol';
import Grid from './Grid';

const stories = storiesOf('Grid', module);

stories
  .addDecorator(withKnobs)
  .addDecorator((storyFn) => (
    <div>
      <GlobalStyles />
      {storyFn()}
    </div>
  ))
  .add('with 12 columns', () => (
    <Grid columns={number('Columns', 12)} gutter={text('Gutter', '1rem')}>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
    </Grid>
  ));
