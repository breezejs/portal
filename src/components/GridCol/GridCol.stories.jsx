import {number, withKnobs} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import React from 'react';
import GlobalStyles from '../GlobalStyles';
import Grid from '../Grid';
import GridCol from './GridCol';

const stories = storiesOf('GridCol', module);

stories
  .addDecorator(withKnobs)
  .addDecorator((storyFn) => (
    <div>
      <GlobalStyles />
      {storyFn()}
    </div>
  ))
  .add('with responsive widths', () => (
    <Grid>
      <GridCol xs={number('xs', 12)} sm={number('sm', 10)} md={number('md', 8)} lg={number('lg', 6)}>Column</GridCol>
    </Grid>
  ));
