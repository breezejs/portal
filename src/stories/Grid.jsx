import {number, text, withKnobs} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import React from 'react';
import GlobalStyles from '../components/GlobalStyles';
import Grid from '../components/Grid';
import GridCol from '../components/GridCol';

const stories = storiesOf('Grid', module);

stories
  .addDecorator(withKnobs)
  .addDecorator((storyFn) => (
    <div>
      <GlobalStyles />
      {storyFn()}
    </div>
  ))
  .add('with 12 columns', () => (
    <Grid columns={number('Columns', 12)} gutter={text('Gutter', '1rem')}>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
      <GridCol xs={1}>col-xs-1</GridCol>
    </Grid>
  ));
