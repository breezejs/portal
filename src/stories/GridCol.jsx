import {number, withKnobs} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import React from 'react';
import GlobalStyles from '../components/GlobalStyles';
import Grid from '../components/Grid';
import GridCol from '../components/GridCol';

const stories = storiesOf('GridCol', module);

stories
  .addDecorator(withKnobs)
  .addDecorator((storyFn) => (
    <div>
      <GlobalStyles />
      {storyFn()}
    </div>
  ))
  .add('with responsive widths', () => (
    <Grid>
      <GridCol xs={number('xs', 12)} sm={number('sm', 10)} md={number('md', 8)} lg={number('lg', 6)}>Column</GridCol>
    </Grid>
  ));
