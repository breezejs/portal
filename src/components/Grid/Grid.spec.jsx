import {mount, shallow} from 'enzyme';
import React from 'react';
import Grid from './Grid';
import GridCol from '../GridCol';

describe('Grid', () => {
  describe('with default values', () => {
    it('should render', () => {
      const result = shallow(
        <Grid />
      );

      expect(result).toMatchSnapshot();
    });

    it('should render with a child components', () => {
      const result = shallow(
        <Grid>
          <GridCol />
        </Grid>
      );

      expect(result).toMatchSnapshot();
    });

    it('should have the correct css styles', () => {
      const result = mount(
        <Grid />
      );

      expect(result.find('div')).toMatchSnapshot();
    });
  });

  describe('with no gutter', () => {
    it('should render', () => {
      const result = shallow(
        <Grid gutter='0' />
      );

      expect(result).toMatchSnapshot();
    });

    it('should render with a child components', () => {
      const result = shallow(
        <Grid gutter='0'>
          <GridCol />
        </Grid>
      );

      expect(result).toMatchSnapshot();
    });

    it('should have the correct css styles', () => {
      const result = mount(
        <Grid gutter='0' />
      );

      expect(result.find('div')).toMatchSnapshot();
    });
  });

  describe('with custom number of columns', () => {
    it('should render', () => {
      const result = shallow(
        <Grid columns={10} />
      );

      expect(result).toMatchSnapshot();
    });

    it('should render with a child components', () => {
      const result = shallow(
        <Grid columns={10}>
          <GridCol />
        </Grid>
      );

      expect(result).toMatchSnapshot();
    });

    it('should have the correct css styles', () => {
      const result = mount(
        <Grid columns={10} />
      );

      expect(result.find('div')).toMatchSnapshot();
    });
  });
});
