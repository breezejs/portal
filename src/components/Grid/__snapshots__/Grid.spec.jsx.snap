// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid with custom number of columns should have the correct css styles 1`] = `
.c0 {
  display: grid;
  grid-template-columns: repeat(10,[col-start] 1fr);
  grid-gap: 1rem;
}

<div
  className="c0"
/>
`;

exports[`Grid with custom number of columns should render 1`] = `
<styled.div
  columns={10}
  gutter="1rem"
/>
`;

exports[`Grid with custom number of columns should render with a child components 1`] = `
<styled.div
  columns={10}
  gutter="1rem"
>
  <Component />
</styled.div>
`;

exports[`Grid with default values should have the correct css styles 1`] = `
.c0 {
  display: grid;
  grid-template-columns: repeat(12,[col-start] 1fr);
  grid-gap: 1rem;
}

<div
  className="c0"
/>
`;

exports[`Grid with default values should render 1`] = `
<styled.div
  columns="12"
  gutter="1rem"
/>
`;

exports[`Grid with default values should render with a child components 1`] = `
<styled.div
  columns="12"
  gutter="1rem"
>
  <Component />
</styled.div>
`;

exports[`Grid with no gutter should have the correct css styles 1`] = `
.c0 {
  display: grid;
  grid-template-columns: repeat(12,[col-start] 1fr);
  grid-gap: 0;
}

<div
  className="c0"
/>
`;

exports[`Grid with no gutter should render 1`] = `
<styled.div
  columns="12"
  gutter="0"
/>
`;

exports[`Grid with no gutter should render with a child components 1`] = `
<styled.div
  columns="12"
  gutter="0"
>
  <Component />
</styled.div>
`;
